partiql-spi/src/main/java/org/partiql/eval/value/DatumComparator.java

```diff
@@ -0,0 +1,720 @@
+package org.partiql.eval.value;
+
+import org.jetbrains.annotations.NotNull;
+import org.partiql.types.PType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * This class allows for the comparison between two {@link Datum}s. This is internally implemented by constructing
+ * a comparison table, where each cell contains a reference to a {@link DatumComparison} to compute the comparison.
+ * The table's rows and columns are indexed by the {@link PType.Kind#ordinal()}. The first dimension matches the
+ * left-hand-side's type of {@link #compare(Datum lhs, Datum rhs)}. The second dimension matches the right-hand-side's
+ * type of {@link #compare(Datum lhs, Datum rhs)}. As such, this implementation allows for O(1) comparison of scalars.
+ */
```

Comment:
Wonder if you had ran any benchmarks comparing the performance between `PartiQLValueComparator` and the `DatumComparator`. If `DatumComparator` is significantly more performant than the previous `PartiQLValueComparator`, it could be a good callout for our upcoming `v1` release.

Reply:
Just finished running a set of benchmarks that shadow this PR and the V1.0.0-perf.1 release. The benchmarks compare evaluation of aggregations, distinct, order by, and comparison operators for the latest engine (this PR), the `v1.0.0-perf.1` engine, and the legacy engine.

See the benchmarks at this [tag](https://github.com/partiql/partiql-lang-kotlin/releases/tag/v1-datum-comparator-benchmark-results.1). See the results [here](https://github.com/partiql/partiql-lang-kotlin/blob/b9d7def1ec4aeab1a93bad69a577ac54efd6dcc7/test/benchmarks/src/jmh/kotlin/org/partiql/benchmarks/tests/EvaluationBenchmarks.kt#L36).

Reply:
And, to hint: it's much faster. In 2 of the 3 benchmarks, the new implementation is over 2x faster than `v1.0.0-perf.1` and over 5x* faster than the legacy implementation.

To get a better overall view of how much the comparator itself is contributing to the performance, it could be a good idea to benchmark AND profile the comparator in action. For now though, we have a good glimpse as for how Datum, PType, and DatumComparator in coordination are improving performance.

The `*` above is discussed in the results link above. The legacy comparator is hard to quantify due to its immediate materialization of aggregations, but preliminary analysis indicates a huge performance improvement nonetheless.

Reply:
Awesome, thanks for creating the benchmarks and tests! The preliminary analysis looks good. Those statements seem like a good starting point for any future performance benchmarking test suite. Ahead of the `v1` release, we should definitely benchmark more queries to show off the performance benefit of the new evaluator.